### PR TITLE
JSDK-2682:  do not retry when connecting 1st time

### PR DESCRIPTION
**Pull request by makarandp0**
With this change we will not retry signaling connection if failed during 1st time. This ensures that we will surface signaling connection error quicker.  
 
I have now added an integration test connects while network is down. It demonstrates the `before` and `after` this change.
 
**It took ~129 seconds to reject before, and ~23 after.**

Before: https://app.circleci.com/jobs/github/twilio/twilio-video.js/10312
```
LOG LOG: 'currentNetworks: ', [Object{Name: 'images_default', Id: '820eb31dff461dae5ef40adc42bbfd6cd1271b7a868e3bfe81d568a49ad5b0fe'}]
WARN LOG: '2020-03-04', '13:40:01.267Z', '|', 'WARN', 'in', '[NTSIceServerSource #3]:', 'Getting ICE servers took too long (using defaults)'
WARN LOG: '2020-03-04', '13:40:21.296Z', '|', 'WARN', 'in', '[TwilioConnection #3: wss://global.vss.twilio.com/signaling]:', 'Closed: 1006 - '
WARN LOG: '2020-03-04', '13:40:41.511Z', '|', 'WARN', 'in', '[TwilioConnection #4: wss://global.vss.twilio.com/signaling]:', 'Closed: 1006 - '
WARN LOG: '2020-03-04', '13:41:01.933Z', '|', 'WARN', 'in', '[TwilioConnection #5: wss://global.vss.twilio.com/signaling]:', 'Closed: 1006 - '
WARN LOG: '2020-03-04', '13:41:22.706Z', '|', 'WARN', 'in', '[TwilioConnection #6: wss://global.vss.twilio.com/signaling]:', 'Closed: 1006 - '
WARN LOG: '2020-03-04', '13:41:44.299Z', '|', 'WARN', 'in', '[TwilioConnection #7: wss://global.vss.twilio.com/signaling]:', 'Closed: 1006 - '
WARN LOG: '2020-03-04', '13:42:07.513Z', '|', 'WARN', 'in', '[TwilioConnection #8: wss://global.vss.twilio.com/signaling]:', 'Closed: 1006 - '
LOG LOG: 'Connect rejected after 129.274 seconds:', 'Signaling connection error'
LOG LOG: 'resetting network'
    ✓ connect rejects when network is down
```

After: https://app.circleci.com/jobs/github/twilio/twilio-video.js/10334
```
LOG LOG: 'currentNetworks: ', [Object{Name: 'images_default', Id: '5ef099cd68fc9c8c7e836088d743e4148726478506c9710bf9d8b1226979d531'}]
WARN LOG: '2020-03-04', '13:40:21.178Z', '|', 'WARN', 'in', '[NTSIceServerSource #3]:', 'Getting ICE servers took too long (using defaults)'
WARN LOG: '2020-03-04', '13:40:41.213Z', '|', 'WARN', 'in', '[TwilioConnection #3: wss://global.vss.twilio.com/signaling]:', 'Closed: 1006 - '
LOG LOG: 'Connect rejected after 23.044 seconds:', 'Signaling connection error'
LOG LOG: 'resetting network'
    ✓ connect rejects when network is down
```

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,16 @@
 For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/support-1.x/CHANGELOG.md).
 
+2.2.1 (In Progress)
```
Comment (manjeshbhargav) [inline CHANGELOG.md]:
```suggestion
2.3.0 (In Progress)
```

File: test/unit/spec/signaling/v2/twilioconnectiontransport.js [hunk 2]
```diff
@@ -1260,6 +1253,10 @@ describe('TwilioConnectionTransport', () => {
             ]);
           });
 
+          it('should set _reconnectAttemptsLeft', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/twilioconnectiontransport.js]: Ideally we should be testing behavior, not private values. The behavior we want to test is that once "connected" is fired on the Transport, an WebSocket disruption should trigger the reconnection logic, and emit "disconnected" on the Transport once all attempts are exhausted.

File: test/unit/spec/signaling/v2/twilioconnectiontransport.js [hunk 3]
```diff
@@ -104,6 +104,10 @@ describe('TwilioConnectionTransport', () => {
         assert.equal('connecting', test.transport.state);
       });
 
+      it('should set the reconnectAttempts to zero initially', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/twilioconnectiontransport.js]: Not necessary to test private values. Ideally we should test the behavior.

Comment (makarandp0) [reply test/unit/spec/signaling/v2/twilioconnectiontransport.js]: @manjeshbhargav - I think in for unit tests it does help with checking private values. and they are cheap.

**Comment by makarandp0** [comment]
@manjeshbhargav - I think updates to the behavior tests cover new scenario nicely. 
```
a "close" event, and the Transport's .state is
        "connected"
          when closed with an Error
            when the re-connect attempts haven't been exhausted
              ✓ should transition .state to "syncing"
            when the re-connect attempts have been exhausted
              ✓ should transition .state to "disconnected"
              ✓ should not emit either "connected" or "message" events
          when closed without an Error
            ✓ should transition .state to "disconnected"
            ✓ should not emit either "connected" or "message" events
        "connecting"
          when closed with an Error
            ✓ should transition .state to "disconnected"
            ✓ should not emit either "connected" or "message" events
          when closed without an Error
            ✓ should transition .state to "disconnected"
            ✓ should not emit either "connected" or "message" events
```

Regarding - checking for the private values - I feel like unit tests gives us this opportunity to inspect things closely, and we should use it. 

about adding a network test using our docker infra - I thought about that, but wasnt sure how do we ensure with a reliable test that 
a) during initial connect we do not attempt multiple times
b) during reconnect we attempt multiple times
since failure is quick, it's not possible to distinguish between the two.  One possibility was to add an extra parameter like waitTimeBetweenAttempts - and with that value large enough we could test such difference reliabally, but that would mean updating the connectOptions. let me know if you have any ideas on how we can write such integration test.

**Comment by makarandp0** [comment]
@manjeshbhargav  - I have added additional unit test that mostly covers existing behavior to ensure that we use reconnect attempts before going into disconnect state.

**Comment by makarandp0** [comment]
@manjeshbhargav, I have added an integration test as we discussed. Can you take another look?

File: test/integration/spec/docker/reconnection.js [hunk 4]
```diff
@@ -212,6 +214,45 @@ describe('Reconnection states and events', function() {
...
+    await waitFor(currentNetworks.map(({ Id: networkId }) => dockerAPI.disconnectFromNetwork(networkId)), 'disconnect from all networks');
+    await waitToGoOffline();
+
+    const start = new Date();
+    let room = null;
+    try {
+      room = await connect(getToken('Alice'), options);
+    } catch (error) {
+      // this exception is expected.
+      const end   = new Date();
```
Comment (manjeshbhargav) [inline test/integration/spec/docker/reconnection.js]:
```suggestion
      const end = new Date();
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,16 @@
...
 
+2.3.0 (In Progress)
+===================
+
+New Features
+------------
+
+- twilio-video.js attempts to connect to signaling server 5 times before giving up. It used to invoke this retry logic
+for initial connection as well as for subsequent re-connections. Now SDK will not attempt retries if connection failed
+during initial connection. This ensures that it surfaces the initial connection errors faster  (JSDK-2682)
```
Comment (manjeshbhargav) [inline CHANGELOG.md]:
```
- twilio-video.js now raises `connect()` errors due to network disruptions quicker by not retrying
  after the first connection attempt fails. (JSDK-2682)
```

**Pull request merged**